src/rust/engine/rule_graph/src/builder.rs

```diff
@@ -1166,13 +1141,70 @@ impl<R: Rule> Builder<R> {
       ))
     } else {
       // Render the most specific errors.
-      Err(Self::render_errored(&graph, errored))
+      Err(Self::render_prune_errors(&graph, errored))
+    }
+  }
+
+  fn render_no_source_of_dependency_error(
+    &self,
+    graph: &MonomorphizedGraph<R>,
+    node: &Node<R>,
+    dependency_key: R::DependencyKey,
+    edge_refs: Vec<EdgeReference<MaybeDeleted<R::DependencyKey, EdgePrunedReason>, u32>>,
+  ) -> String {
+    if self.rules.contains_key(&dependency_key.product()) {
+      format!(
+        "No source of dependency {} for {}. All potential sources were eliminated: {:#?}",
```

Comment:
It's fine to leave for a followup, but this is still fairly esoteric if you don't know how the graph algorithm works. Why were they eliminated? It'd be great to add something like:

```
All potential sources were elimated because they ...

Possible fixes:
* solution1
* solution2
```

Reply:
They each have reasons attached, but yea: it's going to need to be a followup. There is a fair amount of work still to do on error messages unfortunately.